src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Value/EnumConverter.cs

```diff
@@ -601,49 +543,115 @@ private static EnumFieldInfo[] ResolveEnumFields(JsonNamingPolicy? namingPolicy,
             var enumFields = new EnumFieldInfo[names.Length];
             for (int i = 0; i < names.Length; i++)
             {
-                string name = names[i];
+                string originalName = names[i];
                 T value = values[i];
                 ulong key = ConvertToUInt64(value);
-                bool isNameFromAttribute = false;
+                EnumFieldNameKind kind;
 
-                if (enumMemberAttributes != null && enumMemberAttributes.TryGetValue(name, out string? attributeName))
+                if (enumMemberAttributes != null && enumMemberAttributes.TryGetValue(originalName, out string? attributeName))
                 {
-                    name = attributeName;
-                    containsNameAttributes = isNameFromAttribute = true;
+                    originalName = attributeName;
+                    kind = EnumFieldNameKind.Attribute;
+                }
+                else
+                {
+                    kind = namingPolicy != null ? EnumFieldNameKind.NamingPolicy : EnumFieldNameKind.Default;
                 }
 
-                string jsonName = ResolveAndValidateJsonName(name, namingPolicy, isNameFromAttribute);
-                enumFields[i] = new EnumFieldInfo(key, value, name, jsonName, isNameFromAttribute);
+                string jsonName = ResolveAndValidateJsonName(originalName, namingPolicy, kind);
+                enumFields[i] = new EnumFieldInfo(key, kind, originalName, jsonName);
             }
 
             return enumFields;
         }
 
-        private static string ResolveAndValidateJsonName(string name, JsonNamingPolicy? namingPolicy, bool isNameFromAttribute)
+        private static string ResolveAndValidateJsonName(string name, JsonNamingPolicy? namingPolicy, EnumFieldNameKind kind)
         {
-            if (!isNameFromAttribute && namingPolicy is not null)
+            if (kind is not EnumFieldNameKind.Attribute && namingPolicy is not null)
             {
                 // Do not apply a naming policy to names that are explicitly set via attributes.
                 // This is consistent with JsonPropertyNameAttribute semantics.
                 name = namingPolicy.ConvertName(name);
             }
 
-            if (name is null || (s_isFlagsEnum && (name is "" || name.AsSpan().IndexOfAny(' ', ',') >= 0)))
+            if (string.IsNullOrEmpty(name) || char.IsWhiteSpace(name[0]) || char.IsWhiteSpace(name[name.Length - 1]) ||
+                (s_isFlagsEnum && name.AsSpan().IndexOf(',') >= 0))
             {
-                // Reject null strings and in the case of flags additionally reject empty strings or names containing spaces or commas.
+                // Reject null or empty strings or strings with leading or trailing whitespace.
+                // In the case of flags additionally reject strings containing commas.
```

Comment:
It's definitely not motivated by perf, this is a check you only need to make once at startup. It was added originally because F# supports spaces in names but the converter implementation didn't. Given that this now becomes possible in C# via the new attribute, it made sense to just support this scenario for both languages.